src/index.test.ts

```diff
@@ -1,6 +1,1431 @@
+  // atomic, dynamic, and custom types though, so each of these three categories is tested
+  // separately with `null` and `undefined` input. Lastly, there are more tests for various other
+  // edge cases.
+  //
+  // The behavior differs between V3 and V4, so each test has been run for each version. We also
+  // have a block of tests to verify that signatures that match between V3 and V4 remain identical,
+  // and that signatures that differ between V3 and V4 remain different.
+  //
+  // To make reading and maintaining these tests easier, the order will be the same throughout all
+  // 4 of these test suites. Here is a table showing that order, as well as the compatibility of
+  // each input type with V3 and V4 `encodeData`. The table also shows whether the signature is
+  // identical between versions in the cases where the input can be encoded in both versions.
+  //
+  // | Input type                                           | V3 | V4 | Matching Signatures |
+  // | ---------------------------------------------------- | -- | -- | ------------------- |
+  // | Auto-generated tests from the example data           | Y  | Y  | Y                   |
+  // | Arrays using the example data                        | N  | Y  |                     |
+  // | Custom type                                          | Y  | Y  | Y                   |
+  // | Recursive custom type                                | Y  | Y  | N                   |
+  // | Custom type array                                    | N  | Y  |                     |
+  // | Custom type with extra properties                    | Y  | Y  | Y                   |
+  // | Atomic type with `null` input                        | N  | N  |                     |
+  // | Atomic type with `undefined` input                   | Y  | N  |                     |
+  // | Dynamic type with `null` input                       | Y  | Y  | Y                   |
+  // | Dynamic type with `undefined` input                  | Y  | N  |                     |
+  // | Custom type with `null` input                        | N  | Y  |                     |
+  // | Custom type with `undefined` input                   | Y  | Y  | N                   |
+  // | Functions                                            | N  | N  |                     |
+  // | Unrecognized primary type                            | N  | N  |                     |
+  // | Unrecognized non-primary type                        | N  | N  |                     |
+  // | Extra type specified that isn't used by primary type | Y  | Y  | Y                   |
+
+  describe('V3', function () {
+    describe('example data', function () {
+      // Reassigned to silence "no-loop-func" ESLint rule
+      // It was complaining because it saw that `it` and `expect` as "modified variables from the outer scope"
+      // which can be dangerous to reference in a loop. But they aren't modified in this case, just invoked.
+      const _expect = expect;
```

Comment:
I really wish there was some way to silence this false positive. I didn't want to disable the rule for the whole file because it's genuinely useful still, but this is irritating.